### ISSUE TITLE
Assignment 4

### ISSUE THREAD

**Issue by valeriahoyosm** [alexanderquispe/ECO224#1162]
# Instructions

One member of each group has to create a repo on GitHub called **Decision_Trees** to upload this assignment. Then that member should add other group members as collaborators to the repo so that they can all work together in branches. 

There should be a directory for each assignment with the following structure

  ```
 Decision_Trees/
├── Python/
│   ├── input/
│   ├── output/
│   └── scripts/
├── R/
   ├── input/
   ├── output/
   └── scripts/

  ```

To clarify
- In input goes every raw file or dataset your script will use.
- In output goes every file (figures, tables, etc.) your script generates, including a dataset if you are asked to do so.
- In scripts goes your code. You can have multiple scripts organized in some manner by sections, or just have one. 

Your answers to the Comments section should be in .md format and placed in at least one of the three scripts folders for the assignment. To be safe, we advise you to place them in all three. You have until November 2 23:59, to send us a link to the repo by email, indicating your group.

# Assignment 4
Each point is divided by $2$ to grade a specific language. 

## 1. Predicting Heart Disease Using a Classification Tree (10 points)

In this assignment, you will predict whether a person is likely to have heart disease using the dataset located at CausalAI-Course/Data/processed.cleveland.data.

You will use a classification tree model. Set random_state = 123 for both the classification tree and the cross-validation folds.

## 1.1 Data Cleaning (2 points)

- Rename the variables in the following order:
['age', 'sex', 'cp', 'restbp', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'hd'],
and remove missing values.
- Convert all categorical variables into dummy variables.
- Create a binary variable y that equals 1 if the person has heart disease and 0 otherwise.

## 1.2 Data Analysis (8 points)

- (1 point) Split the data into training and test sets, and plot a classification tree (random_state = 123).
- (2 points) Plot the confusion matrix using the labels ["Does not have HD", "Has HD"] and interpret the results.
- (1.5 points) Fix the overfitting problem using ross-validation.
Generate 50 values of α equally spaced on a logarithmic scale between e⁻¹⁰ and 0.05.
Use 4-fold cross-validation to select the optimal alpha (random_state = 123).
- (1.5 points) Plot the Inaccuracy Rate (1 − Accuracy) against alpha.
- (2 points)  Plot again the classification tree and the confusion matrix using the optimal α. Interpret and briefly discuss the results.

## 2. Causal Forest (10 points)

We will continue using the dataset from the previous section.
Suppose a random cash transfer program is implemented to encourage individuals to attend a medical check-up for potential heart problems.

- (0.5 points) Create a binary treatment variable T that randomly takes values 0 or 1 with equal probability (simulating random assignment to the program).
- (1 point) Create an outcome variable Y that measures the improvement in cardiovascular health after the program, following the rule below:

$$
Y = (1 + 0.05 *age + 0.3*sex + 0.2 *restbp)*T + 0.5*oldpeak  \epsilon  
$$

$$
\quad \epsilon \sim N(0,1)
$$

- (1 point) Calculate the treatment effect on your new outcome variable 𝑌 using OLS.

- (2 points) Use a Random Forest model to estimate the causal effects.

- (2 points) Plot a representative tree that captures the most important heterogeneous treatment effects with a maximum depth of 2 (max_depth=2) and interpret it.

- (1.5 points) Compute the feature importances and visualize them in a bar chart.

- (2 points) Plot the distribution of standardized covariates by predicted treatment effect terciles:
    - Standardize all covariates.

    - Divide the predicted treatment effect (or your outcome variable if no predicted effect is available) into three terciles: low, medium, high.

    - Compute the mean of each covariate within each tercile.

    - Create a heatmap showing these means, with covariates on the x-axis, terciles on the y-axis, and color representing the mean value.

| Variable  | Description |
|------------|-------------|
| **age** | Age of the individual |
| **sex** | Sex (1 = male, 0 = female) |
| **cp** | Chest pain type |
| **restbp** | Resting blood pressure (in mm Hg) |
| **chol** | Serum cholesterol (in mg/dl) |
| **fbs** | Fasting blood sugar (> 120 mg/dl; 1 = true, 0 = false) |
| **restecg** | Resting electrocardiographic results |
| **thalach** | Maximum heart rate achieved |
| **exang** | Exercise-induced angina |
| **oldpeak** | ST depression induced by exercise relative to rest |
| **slope** | The slope of the peak exercise ST segment |
| **ca** | Number of major vessels (0–3) colored by fluoroscopy |
| **thal** | Result of thallium heart scan (3 = normal, 6 = fixed defect, 7 = reversible defect) |
| **hd** | Diagnosis of heart disease (the predicted attribute) |

Deadline = November 2 23:59.
